File: src/utils/rng.js
// Small deterministic PRNG (mulberry32) so a given seed always produces the
// same world. Also exposes a shared "loose" RNG for gameplay randomness that
// doesn't need to be reproducible.

export function mulberry32(seed) {
  let a = seed >>> 0;
  return function rand() {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function makeRng(seed) {
  const rand = mulberry32(seed);
  return {
    next: rand,
    range(min, max) { return min + rand() * (max - min); },
    int(min, max) { return Math.floor(min + rand() * (max - min + 1)); },
    chance(p) { return rand() < p; },
    pick(arr) { return arr[Math.floor(rand() * arr.length)]; },
  };
}

export const rng = makeRng(Date.now() & 0xffffffff);

File: src/main/java/com/worldbox/util/Rng.java
package com.worldbox.util;

import java.util.List;

/** Deterministic mulberry32 PRNG so a given seed always produces the same
 * world; also handy as a general-purpose gameplay RNG. Java's int math
 * already wraps to 32 bits, so this is a direct port of the JS version
 * without needing an explicit Math.imul-style helper. */
public class Rng implements java.io.Serializable {
  private int a;

  public Rng(long seed) {
    this.a = (int) seed;
  }

  private double rawDouble() {
    a += 0x6d2b79f5;
    int t = a;
    t = (t ^ (t >>> 15)) * (t | 1);
    t = (t + (t ^ (t >>> 7)) * (t | 61)) ^ t;
    int result = t ^ (t >>> 14);
    return (result & 0xFFFFFFFFL) / 4294967296.0;
  }

  public double next() { return rawDouble(); }
  public double range(double min, double max) { return min + rawDouble() * (max - min); }
  public int intRange(int min, int max) { return (int) Math.floor(min + rawDouble() * (max - min + 1)); }
  public boolean chance(double p) { return rawDouble() < p; }
  public <T> T pick(List<T> list) { return list.get((int) Math.floor(rawDouble() * list.size())); }
  public <T> T pick(T[] arr) { return arr[(int) Math.floor(rawDouble() * arr.length)]; }
}
